### DISCUSSION TITLE
Set field on readonly based on a block - Unable to save due to unpermitted_params

### DISCUSSION THREAD

**Discussion by lpe** [avo-hq/avo#1458] [Help]
This is my code in the resource :
`field :accountnr, as: :text, name: "Code comptable", readonly: -> { !current_user.compta? }`

This is the error when I try to edit / save :
`method=PUT path=/resources/accounts/22772 format=turbo_stream controller=Avo::AccountsController action=update status=302 duration=27.26 view=0.00 db=4.79 location=http://localhost:3000/resources/accounts/22772 unpermitted_params=["_method", "authenticity_token", "account", "resellers-dummy", "categories-dummy", "fake_lastcontact", "button", "accountnr"] ...`

It seems this is due to the block as this is working :
`field :accountnr, as: :text, name: "Code comptable", readonly: false`

This is working on another field :
`field :actif, as: :boolean, index_text_align: :center, hide_on: [:show], readonly: -> { !current_user.compta? }`

I don't understand why ...

Updated to 2.19, same issue.

Any idea ? Thx by advance.

**Comment by adrianthedev** [comment]
So what is the error you are getting @lpe? Is the validation failing because the field is not being sent in?

**Reply by adrianthedev** [reply]
The read-only field works like this:

- the field is shown as disabled on the page
- this is because a bad actor might go into the inspect element panel and re-enable the disabled field and send in bad data
- when the form is submitted, the value sent by the form is not assigned to the model

Do you think this is why the form submission is failing?

**Reply by lpe** [reply]
Not really a validation error.  The field value is not transmitted in the params :
` Account Update (0.5ms)  UPDATE "accounts" SET "updated_at" = $1, "pays" = $2 WHERE "accounts"."id" = $3  [["updated_at", "2022-11-28 14:47:43.330276"], ["pays", "BELGIUMX"], ["id", 22772]]`
Seems it's due to the block `readonly: -> { !current_user.compta? }` as `readonly: false` is working
Strange that it works on another field : `field :actif, as: :boolean, index_text_align: :center, hide_on: [:show], readonly: -> { !current_user.compta? }`  The only difference is `hide_on: [:show]`
Perhaps is therre a better approach ? 
Thx by advance for your help.

**Reply by adrianthedev** [reply]
So you're saying that if `!current_user.compta?` evaluates to `false`, Avo does something different than when you set it manually to `false`?

Ex: `readonly: -> { !current_user.compta? # false } `, is different than `readonly: false`.

**Reply by lpe** [reply]
Something else weird : `visible: -> (resource:) {resource.current_user.compta? } ` is working, not `readonly: -> (resource:) {resource.current_user.compta? } ` Error : `missing keyword: :resource`

**Comment by lpe** [comment]
Yes. Please note that I dit not set a context in the avo.rb intializer as `config.current_user_method = :current_user` is already set.

**Reply by adrianthedev** [reply]
ok. so what does `{ !current_user.compta? }` evaluate to? `true`? Because if so, then it means it works ok. that parameter is not being sent to the server as a security measure. Read [my initial message](https://github.com/avo-hq/avo/discussions/1458#discussioncomment-4253333).

What you might be able to do is to add a hidden field below that one and have that value being sent to the server.

```ruby
field :accountnr, as: :text, name: "Code comptable", readonly: -> { !current_user.compta? }
field :accountnr, as: :hidden

```

**Reply by lpe** [reply]
Thx for your help. This does not work :-/

So now I have : 
```ruby
field :accountnr, as: :text, name: "Code comptable", readonly: -> { !current_user.compta? }
field :accountnr, as: :hidden
```

accountnr is clearly notr transmitted (below the log where I have changed 2 fields (accountnr and pays) :

```
11:46:24 web.1  | Account Update (2.9ms)  UPDATE "accounts" SET "updated_at" = $1, "pays" = $2 WHERE "accounts"."id" = $3  [["updated_at", "2022-12-01 10:46:24.611244"], ["pays", "BELGIUMX"], ["id", 22772]]
11:46:24 web.1  | method=PUT path=/resources/accounts/22772 format=turbo_stream controller=Avo::AccountsController action=update status=302 duration=57.72 view=0.00 db=31.67 location=http://localhost:3000/resources/accounts/22772 unpermitted_params=["_method", "authenticity_token", "account", "resellers-dummy", "categories-dummy", "fake_lastcontact", "button"] time=2022-12-01 11:46:24 +0100
```

**Reply by adrianthedev** [reply]
OK. Can you please try one more thing? keep the `hidden` field and remove the `text` one and see it it updates the record.

```ruby
field :accountnr, as: :hidden
```

**Reply by adrianthedev** [reply]
And if that doesn't work, try it without the `readonly` flag.

```ruby
field :accountnr, as: :text, name: "Code computable"
```

I want to make sure it's Avo and not something from your app.

**Reply by lpe** [reply]
Yes, this is working and `readonly: false ` too.

What do you think of `action=update status=302` in the log ? And why we have an error on `readonly: -> (resource:) {resource.current_user.compta? }` `Error : missing keyword: :resource` ?

**Reply by adrianthedev** [reply]
Hey @lpe, are you available for a quick video call to show me your setup?
https://whereby.com/adrianthedev

**Reply by adrianthedev** [reply]
I'm off the call link, but try to use this version `2.21.pre.1.issue1458` (`gem "avo", "2.21.pre.1.issue1458") and instead of `readonly: -> { !current_user.compta? }` use `disabled: -> { !current_user.compta? }`.

**Reply by lpe** [reply]
Sorry, I was not in dev at this time. But good news : this is working, event without the hidden property.

**Reply by lpe** [reply]
Unfortunately I am not able anymore to edit the related contracts (account has_many contracts /contract belongs_to contract

`undefined method  'disabled?' for #<Avo::Fields::BelongsToField`

**Reply by lpe** [reply]
It's really buggy. I removed the readonly and disabled attributes because I can't modify the linked resources anymore. This is not fatal at this stage of the project. I will follow the updates and come back to the stable version.

**Reply by adrianthedev** [reply]
yeah. that's not polished yet. I just wanted to see if that fixes your exact issue.

the `disavbled` option will be better in the next release (next Tuesday)
